File: docx-formatter-plus/scripts/formulas.js
/**
 * formulas.js
 * Formula definition template using mathHelpers.js.
 *
 * Each formula is a DMath object that can be placed in a centered Paragraph.
 * Import: const { eq1, eq2, eq3, eq4 } = require("./formulas");
 *
 * Add new formulas following the same pattern.
 * Greek letters: use Unicode directly (λ=\u03bb, τ=\u03c4, α=\u03b1, β=\u03b2, Σ=\u2211)
 */

const { r, sub, sup, frac, sumOp, func, paren, bracket, math } = require("./mathHelpers");

// ============================================================
// Formula 1: L_total = L_LLM + λ × L_ontology
// ============================================================
const eq1 = math([
  sub("L", "total"),
  r(" = "),
  sub("L", "LLM"),
  r(" + \u03bb \u00d7 "),  // λ ×
  sub("L", "ontology"),
]);

// ============================================================
// Formula 2: L_LLM = −Σ_i y_i log(p_i)  (cross-entropy loss)
// ============================================================
const eq2 = math([
  sub("L", "LLM"),
  r(" = \u2212"),  // −
  sumOp([r("i")], [
    sub("y", "i"),
    r(" "),
    func("log", [sub("p", "i")]),
  ]),
]);

// ============================================================
// Formula 3: L_ontology (InfoNCE contrastive loss)
// L_ontology = −Σ_i [ (1/|P(i)|) Σ_{j∈P(i)} log( exp(sim(e_i,e_j)/τ) / [exp(sim(e_i,e_j)/τ) + Σ_{k∈N(i)} w_k · exp(sim(e_i,e_k)/τ)] ) ]
// ============================================================

// Helper: sim(e_i, e_j) / τ as a fraction
function simFrac(iSub, jSub) {
  return frac(
    [r("sim("), sub("e", iSub), r(","), sub("e", jSub), r(")")],
    [r("\u03c4")]  // τ
  );
}

// Helper: exp(sim(e_i, e_j) / τ)
function expSim(iSub, jSub) {
  return func("exp", [simFrac(iSub, jSub)]);
}

// Denominator: exp(sim(e_i,e_j)/τ) + Σ_{k∈N(i)} w_k · exp(sim(e_i,e_k)/τ)
const denomInner = [
  expSim("i", "j"),
  r(" + "),
  sumOp([r("k\u2208N(i)")], [  // k∈N(i)
    sub("w", "k"),
    r(" \u00b7 "),  // ·
    expSim("i", "k"),
  ]),
];

// Big fraction: exp(sim(e_i,e_j)/τ) / denominator
const bigFraction = frac([expSim("i", "j")], denomInner);

// log(bigFraction)
const logTerm = func("log", [bigFraction]);

// Inner sum: Σ_{j∈P(i)} log(...)
const innerSum = sumOp([r("j\u2208P(i)")], [logTerm]);

// Coefficient: 1 / |P(i)|
const coeff = frac([r("1")], [r("|P(i)|")]);

// Outer sum: Σ_i [ coeff × innerSum ]
const outerSum = sumOp([r("i")], [
  bracket([coeff, r(" "), innerSum]),
]);

const eq3 = math([
  sub("L", "ontology"),
  r(" = \u2212"),  // −
  outerSum,
]);

// ============================================================
// Formula 4: W = W₀ + B × A  (LoRA decomposition, r ≪ d)
// ============================================================
const eq4 = math([
  r("W"),
  r(" = "),
  sub("W", "0"),
  r(" + B \u00d7 A"),  // ×
]);

module.exports = { eq1, eq2, eq3, eq4 };

File: scripts/formulas.py
# -*- coding: utf-8 -*-
"""
formulas.py
Formula definitions using mathHelpers.py — 1:1 port of formulas.js.

Each formula is an OMML XML string that can be placed via add_eq_para(doc, eqN).
Import: from formulas import eq1, eq2, eq3, eq4

Add new formulas following the same pattern.
Greek letters: use Unicode directly (λ=\\u03bb, τ=\\u03c4, α=\\u03b1, β=\\u03b2, Σ=\\u2211)
"""

from mathHelpers import r, sub, sup, frac, sumOp, func, paren, bracket, math

# ============================================================
# Formula 1: L_total = L_LLM + λ × L_ontology
# ============================================================
eq1 = math([
    sub("L", "total"),
    r(" = "),
    sub("L", "LLM"),
    r(" + \u03bb \u00d7 "),  # λ ×
    sub("L", "ontology"),
])

# ============================================================
# Formula 2: L_LLM = −Σ_i y_i log(p_i)  (cross-entropy loss)
# ============================================================
eq2 = math([
    sub("L", "LLM"),
    r(" = \u2212"),  # −
    sumOp([r("i")], [
        sub("y", "i"),
        r(" "),
        func("log", [sub("p", "i")]),
    ]),
])

# ============================================================
# Formula 3: L_ontology (InfoNCE contrastive loss)
# L_ontology = −Σ_i [ (1/|P(i)|) Σ_{j∈P(i)} log( exp(sim(e_i,e_j)/τ) / [exp(sim(e_i,e_j)/τ) + Σ_{k∈N(i)} w_k · exp(sim(e_i,e_k)/τ)] ) ]
# ============================================================


def _sim_frac(i_sub, j_sub):
    """Helper: sim(e_i, e_j) / τ as a fraction."""
    return frac(
        [r("sim("), sub("e", i_sub), r(","), sub("e", j_sub), r(")")],
        [r("\u03c4")],  # τ
    )


def _exp_sim(i_sub, j_sub):
    """Helper: exp(sim(e_i, e_j) / τ)."""
    return func("exp", [_sim_frac(i_sub, j_sub)])


# Denominator: exp(sim(e_i,e_j)/τ) + Σ_{k∈N(i)} w_k · exp(sim(e_i,e_k)/τ)
_denom_inner = [
    _exp_sim("i", "j"),
    r(" + "),
    sumOp([r("k\u2208N(i)")], [  # k∈N(i)
        sub("w", "k"),
        r(" \u00b7 "),  # ·
        _exp_sim("i", "k"),
    ]),
]

# Big fraction: exp(sim(e_i,e_j)/τ) / denominator
_big_fraction = frac([_exp_sim("i", "j")], _denom_inner)

# log(bigFraction)
_log_term = func("log", [_big_fraction])

# Inner sum: Σ_{j∈P(i)} log(...)
_inner_sum = sumOp([r("j\u2208P(i)")], [_log_term])

# Coefficient: 1 / |P(i)|
_coeff = frac([r("1")], [r("|P(i)|")])

# Outer sum: Σ_i [ coeff × innerSum ]
_outer_sum = sumOp([r("i")], [
    bracket([_coeff, r(" "), _inner_sum]),
])

eq3 = math([
    sub("L", "ontology"),
    r(" = \u2212"),  # −
    _outer_sum,
])

# ============================================================
# Formula 4: W = W₀ + B × A  (LoRA decomposition, r ≪ d)
# ============================================================
eq4 = math([
    r("W"),
    r(" = "),
    sub("W", "0"),
    r(" + B \u00d7 A"),  # ×
])
